File: main-service/src/main/java/ru/pgk/main_service/features/teacher/controllers/TeacherController.java
package ru.pgk.main_service.features.teacher.controllers;

import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.security.SecurityRequirements;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.web.bind.annotation.*;
import ru.pgk.main_service.common.dto.PageDto;
import ru.pgk.main_service.features.teacher.dto.TeacherDetailsDto;
import ru.pgk.main_service.features.teacher.dto.params.AddOrUpdateTeacherParams;
import ru.pgk.main_service.features.teacher.mappers.TeacherDetailsMapper;
import ru.pgk.main_service.features.teacher.service.TeacherService;
import ru.pgk.main_service.features.teacher.service.queries.TeacherQueriesService;

@RestController
@RequestMapping("teachers")
@RequiredArgsConstructor
public class TeacherController {

    private final TeacherService teacherService;
    private final TeacherQueriesService teacherQueriesService;

    private final TeacherDetailsMapper teacherDetailsMapper;

    @GetMapping
    private PageDto<TeacherDetailsDto> getAll(
            @RequestParam(required = false) String name,
            @RequestParam(defaultValue = "0") Integer offset
    ) {
        Page<TeacherDetailsDto> teachers = teacherQueriesService.getAll(name, offset).map(teacherDetailsMapper::toDto);
        return PageDto.fromPage(teachers);
    }

    @PostMapping
    @SecurityRequirements(
            value = {
                    @SecurityRequirement(name = "bearerAuth"),
                    @SecurityRequirement(name = "X-API-KEY")
            }
    )
    private TeacherDetailsDto add(
            @RequestBody AddOrUpdateTeacherParams params
    ) {
        return teacherDetailsMapper.toDto(teacherService.add(params));
    }

    @PutMapping("{id}")
    @SecurityRequirements(
            value = {
                    @SecurityRequirement(name = "bearerAuth"),
                    @SecurityRequirement(name = "X-API-KEY")
            }
    )
    private TeacherDetailsDto update(
            @PathVariable Integer id,
            @RequestBody AddOrUpdateTeacherParams params
    ) {
        return teacherDetailsMapper.toDto(teacherService.update(id, params));
    }

    @DeleteMapping("{id}")
    @SecurityRequirements(
            value = {
                    @SecurityRequirement(name = "bearerAuth"),
                    @SecurityRequirement(name = "X-API-KEY")
            }
    )
    private void deleteById(
            @PathVariable Integer id
    ) {
        teacherService.deleteById(id);
    }
}

File: teacher-service/src/main/kotlin/ru/pgk/teacher_service/features/controllers/TeacherController.kt
package ru.pgk.teacher_service.features.controllers

import org.springframework.web.bind.annotation.*
import ru.pgk.teacher_service.common.dto.PageDto
import ru.pgk.teacher_service.common.dto.toDto
import ru.pgk.teacher_service.features.dto.TeacherDetailsDto
import ru.pgk.teacher_service.features.dto.params.AddOrUpdateTeacherParams
import ru.pgk.teacher_service.features.mappers.TeacherDetailsMapper
import ru.pgk.teacher_service.features.services.TeacherService
import ru.pgk.teacher_service.features.services.queries.TeacherQueriesService

@RestController
class TeacherController(
    private val teacherQueriesService: TeacherQueriesService,
    private val teacherService: TeacherService,

    private val teacherDetailsMapper: TeacherDetailsMapper
) {

    @GetMapping
    private fun getAll(
        @RequestParam(required = false) name: String?,
        @RequestParam(defaultValue = "0") offset: Int
    ): PageDto<TeacherDetailsDto> {
        val teachers = teacherQueriesService.getAll(name, offset).map(teacherDetailsMapper::toDto)
        return teachers.toDto()
    }

    @PostMapping
    private fun add(
        @RequestBody params: AddOrUpdateTeacherParams,
    ): TeacherDetailsDto {
        return teacherDetailsMapper.toDto(teacherService.add(params))
    }

    @PutMapping("{id}")
    private fun update(
        @PathVariable id: Int,
        @RequestBody params: AddOrUpdateTeacherParams
    ): TeacherDetailsDto {
        return teacherDetailsMapper.toDto(teacherService.update(id, params))
    }

    @DeleteMapping("{id}")
    private fun deleteById(
        @PathVariable id: Int,
    ) {
        teacherService.deleteById(id)
    }
}